src/java/org/apache/cassandra/index/sai/metrics/AbstractMetrics.java

```diff
@@ -57,28 +55,28 @@ public abstract class AbstractMetrics
 
     public void release()
     {
-        tracked.forEach(Metrics::remove);
-        tracked.clear();
+        Metrics.removeIfMatch(fullName -> resolveShortMetricName(fullName,  DefaultNameFactory.GROUP_NAME, TYPE),
+                              this::createMetricName, m -> {});
     }
 
-    protected CassandraMetricsRegistry.MetricName createMetricName(String name)
+    protected String metricScope()
     {
-        return createMetricName(name, scope);
+        return scope;
     }
 
-    protected CassandraMetricsRegistry.MetricName createMetricName(String name, String scope)
+    protected CassandraMetricsRegistry.MetricName createMetricName(String name)
     {
-        String metricScope = keyspace + '.' + table;
-        if (index != null)
-        {
-            metricScope += '.' + index;
-        }
-        metricScope += '.' + scope + '.' + name;
+        return createMetricName(name, metricScope());
+    }
 
-        CassandraMetricsRegistry.MetricName metricName = new CassandraMetricsRegistry.MetricName(DefaultNameFactory.GROUP_NAME,
-                                                                                                 TYPE, name, metricScope, createMBeanName(name, scope));
-        tracked.add(metricName.getMetricName());
-        return metricName;
+    private CassandraMetricsRegistry.MetricName createMetricName(String name, String scope)
+    {
+        assert name.indexOf('.') == -1 : String.format("Metric name '%s' should not contain '.'", name);
+        return new CassandraMetricsRegistry.MetricName(DefaultNameFactory.GROUP_NAME,
+                                                       TYPE,
+                                                       name,
+                                                       MetricRegistry.name(keyspace, table, index, scope, name),
```

Comment:
It feels like something around this change is causing problems w/ `index` handling that pop up in `GuardrailNonPartitionRestrictedQueryTest`, `IndexAvailabilityTest`, and a few other places.

Reply:
Some of the metrics are created with `scope = null`, so the `resolveShortMetricName` method won't work in this case. This is an implementation bug. Fixed.